marshal.go

```diff
@@ -0,0 +1,279 @@
+	// TODO allow body types that aren't necessarily JSON.
+	bodyValue := dereferenceIfPointer(v)
+	if bodyValue == emptyValue {
+		return nil
+	}
+
+	if p.Method != "POST" && p.Method != "PUT" {
+		return errgo.Newf("trying to marshal to body of a request with method %q", p.Method)
+	}
+
+	data, err := json.Marshal(bodyValue.Interface())
+	if err != nil {
+		return errgo.Notef(err, "cannot marshal request body")
+	}
+	p.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+	return nil
+}
+
+// marshalAllField marshals a []string slice into form fields.
+func marshalAllField(name string) marshaler {
+	return func(v reflect.Value, p *Params) error {
+		value := dereferenceIfPointer(v)
+		if value == emptyValue {
+			return nil
+		}
+		values := value.Interface().([]string)
+		if p.Form == nil {
+			p.Form = url.Values{}
+		}
+		for _, value := range values {
+			p.Form.Add(name, value)
+		}
+		return nil
+	}
+}
+
+// marshalString marshals s string field.
+func marshalString(tag tag) marshaler {
+	formSet := formSetters[tag.source]
+	if formSet == nil {
+		panic("unexpected source")
+	}
+	return func(v reflect.Value, p *Params) error {
+		value := dereferenceIfPointer(v)
+		if value == emptyValue {
+			return nil
+		}
+		formSet(tag.name, value.String(), p)
+		return nil
+	}
+}
+
+var textMarshalerType = reflect.TypeOf((*encoding.TextMarshaler)(nil)).Elem()
+
+func implementsTextMarshaler(t reflect.Type) bool {
+	// Use the pointer type, because a pointer
+	// type will implement a superset of the methods
+	// of a non-pointer type.
+	return reflect.PtrTo(t).Implements(textMarshalerType)
+}
+
+// marshalWithMarshalText returns a marshaler
+// that marshals the given type from the given tag
+// using its MarshalText method.
+func marshalWithMarshalText(t reflect.Type, tag tag) marshaler {
+	formSet := formSetters[tag.source]
+	if formSet == nil {
+		panic("unexpected source")
+	}
+	return func(v reflect.Value, p *Params) error {
+		value := dereferenceIfPointer(v)
+		if value == emptyValue {
+			return nil
+		}
+		m := value.Addr().Interface().(encoding.TextMarshaler)
+		data, err := m.MarshalText()
+		if err != nil {
+			return errgo.Mask(err)
+		}
+		formSet(tag.name, string(data), p)
+
+		return nil
+	}
+}
+
+// marshalWithScan returns an marshaler
+// that unmarshals the given tag using fmt.Sprint.
+func marshalWithSprint(tag tag) marshaler {
+	formSet := formSetters[tag.source]
+	if formSet == nil {
+		panic("unexpected source")
+	}
+	return func(v reflect.Value, p *Params) error {
+		value := dereferenceIfPointer(v)
+		if value == emptyValue {
+			return nil
+		}
+		valueString := fmt.Sprint(value.Interface())
+
+		formSet(tag.name, valueString, p)
+
+		return nil
+	}
+}
+
+// formSetters maps from source to a function that
+// sets the value for a given key.
+var formSetters = []func(string, string, *Params){
+	sourceForm: func(name, value string, p *Params) {
+		if p.Form == nil {
+			p.Form = url.Values{}
+		}
+		p.Form.Add(name, value)
```

Comment:
I think this should probably be Set. If people have several fields that specify the same form field,
then they will have a problem - even if we marshal all the values into the same field here,
they will unmarshal only the first item into all the values.

We should probably check in parseType that no field is specified more than once - something
for a future PR, I think.